blockstore.go

```diff
@@ -1,15 +1,65 @@
 package main
 
 import (
+	"context"
 	"crypto/tls"
 	"net/http"
 	"net/url"
 	"time"
 
 	"github.com/filecoin-saturn/caboose"
+	"github.com/ipfs/go-cid"
 	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	exchange "github.com/ipfs/go-ipfs-exchange-interface"
+	blocks "github.com/ipfs/go-libipfs/blocks"
 )
 
+func newExchange(orchestrator, loggingEndpoint string) (exchange.Interface, error) {
+	b, err := newBlockStore(orchestrator, loggingEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	return &exchangeBsWrapper{bstore: b}, nil
+}
+
+type exchangeBsWrapper struct {
+	bstore blockstore.Blockstore
+}
+
+func (e *exchangeBsWrapper) GetBlock(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+
+	return e.bstore.Get(ctx, c)
+}
+
+func (e *exchangeBsWrapper) GetBlocks(ctx context.Context, cids []cid.Cid) (<-chan blocks.Block, error) {
+	out := make(chan blocks.Block)
+
+	go func() {
+		defer close(out)
+		for _, c := range cids {
+			blk, err := e.GetBlock(ctx, c)
+			if err != nil {
+				return
+			}
+			out <- blk
+		}
+	}()
+	return out, nil
+}
+
+func (e *exchangeBsWrapper) NotifyNewBlocks(ctx context.Context, blks ...blocks.Block) error {
+	//TODO implement me
+	return nil
+}
```

Comment:
> tried to understand the purpose of NotifyNewBlocks but docs/examples are scarce. Is it used only in the context of making blocks added to local blockstore via ipfs add available to ongoing bitswap sessions(?)

It should be fine to just return nil here.

IIUC `NotifyNewBlocks` has two points:
1. Let the server know it has blocks that are available to be served (i.e. because if you ask a Bitswap server for a block and it later receives it it can send it to you without you asking again). These blocks could come from an exchange fetch (i.e. go-bitswap does not do Puts into the blockstore itself that happens in go-blockservice) or from an external source (e.g. `ipfs add`).
2. Let the client know that a block has arrived that it was otherwise waiting for
   - This doesn't really need to exist, but if the exchange was waiting for a block that came in from somewhere else (e.g. a CAR file import) then the call would either need to be cancelled + handled by a higher layer, like the blockservice, or the block would need to get threaded through the call stack. This enables the latter choice.
   - I suspect this case being covered is more an accident of needing to handle the first one and go-bitswap historically combining client and server elements

Reply:
Note: The reason we can return `nil` here isn't so much that it's a good idea or anything but that this whole wrapper shouldn't exist as caboose should just be an exchange. In that context, caboose would need to decide whether it wants to be smart enough to deal with what happens if you ask for the same block twice close together.

My suspicion is that they won't care because further down the road we're going to be dealing with graphs rather than blocks here anyway.